blockchain/indexers/common.go

```diff
@@ -173,3 +171,788 @@ func interruptRequested(ctx context.Context) bool {
+	// disconnected from the main chain.
+	DisconnectNtfn
+)
+
+// IndexNtfn represents an index notification detailing a block connecting
+// or disconnecting from the main chain.
+type IndexNtfn struct {
+	NtfnType    IndexNtfnType
+	Block       *dcrutil.Block
+	Parent      *dcrutil.Block
+	PrevScripts PrevScripter
+}
+
+// IndexSubscription represents a subscription for index updates.
+type IndexSubscription struct {
+	id         string
+	idx        Indexer
+	subscriber *IndexSubscriber
+	c          chan *IndexNtfn
+	mtx        sync.Mutex
+
+	// prerequisite defines the notification processing hierarchy for this
+	// subscription. It is expected that the subscriber associated with the
+	// prerequisite provided processes notifications before they are
+	// delivered by this subscription to its subscriber. An empty string
+	// indicates the subscription has no prerequisite.
+	prerequisite string
+
+	// dependency defines the index subscription that requires the subscriber
+	// associated with this subscription to have processed incoming
+	// notifications before it does. A nil dependency indicates the subscription
+	// has no dependencies.
+	dependency *IndexSubscription
+}
+
+// newIndexSubscription initializes a new index subscription.
+func newIndexSubscription(subber *IndexSubscriber, index Indexer, prereq string) *IndexSubscription {
+	return &IndexSubscription{
+		id:           index.Name(),
+		idx:          index,
+		prerequisite: prereq,
+		subscriber:   subber,
+		c:            make(chan *IndexNtfn, bufferSize),
+	}
+}
+
+// C returns a channel that produces a stream of index notifications.
+// Successive calls to C return the same channel.
+func (s *IndexSubscription) C() chan *IndexNtfn {
+	return s.c
+}
+
+// Stop prevents any future index updates from being delivered and
+// unsubscribes the associated subscription.
+//
+// The subscriber mutex must be held when this function is called.
+func (s *IndexSubscription) Stop() error {
+	// If the subscription has a prerequisite, find it and remove the
+	// subscription as a dependency.
+	if s.prerequisite != noPrereqs {
+		prereq, ok := s.subscriber.subscriptions[s.prerequisite]
+		if !ok {
+			return fmt.Errorf("no subscription found with id %s", s.prerequisite)
+		}
+
+		prereq.mtx.Lock()
+		prereq.dependency = nil
+		prereq.mtx.Unlock()
+
+		return nil
+	}
+
+	// If the subscription has a dependency, stop it as well.
+	if s.dependency != nil {
+		err := s.dependency.Stop()
+		if err != nil {
+			return err
+		}
+	}
+
+	// If the subscription is independent, remove it from the
+	// index subscriber's subscriptions.
+	delete(s.subscriber.subscriptions, s.id)
+
+	return nil
+}
+
+// publishIndexNtfn sends the provided index notification on the channel
+// associated with the subscription.
+func (s *IndexSubscription) publishIndexNtfn(ntfn *IndexNtfn) {
+	select {
+	case s.c <- ntfn:
+	default:
+		// Non-blocking send fallthrough.
+	}
```

Comment:
This causes notifications to be missed by the subscribers when the buffer is full, and therefore the index will be missing data.  I hit this case even with the buffer size of 128, but you can more easily simulate this by changing the buffer size to 1.  If you then attempt to query for an address/tx within that missed block, you'll see that the data is missing, even though the data may be there for later blocks.

The current assumption/design is the blocks are connected to the indexers in order (since, for example, if the tip height is 100, the indexer is assuming that all blocks up to 100 were connected).  In that case, you may want to error if a block is connected and the height is not the current tip height plus one so that this wouldn't be silently missed.

Unless you go with an alternative design, you'll need to block when the buffer is full here since the indexers currently rely on the notifications for each block.